File: scripts/generate-qr-pdf.py
#!/usr/bin/env python3
"""
Générateur de PDF QR Codes - GXO Moissy
Génère un PDF avec tous les QR Codes des 45 quais
"""

import qrcode
from reportlab.lib.pagesizes import A4
from reportlab.lib.units import mm
from reportlab.pdfgen import canvas
from reportlab.lib import colors
from datetime import datetime
import os

# Configuration des quais
QUAIS_CONFIG = {
    'Zone A': list(range(1, 11)),  # 1-10
    'Zone B': list(range(32, 39)),  # 32-38
    'Zone C': list(range(45, 50)),  # 45-49
    'Zone D': [60, 61, 62, 67, 68, 69],
    'Zone E': list(range(75, 80)) + list(range(81, 88)),  # 75-79, 81-87
    'Zone F': list(range(99, 104))  # 99-103
}

ZONE_COLORS = {
    'Zone A': (59/255, 130/255, 246/255),   # Bleu
    'Zone B': (139/255, 92/255, 246/255),   # Violet
    'Zone C': (245/255, 158/255, 11/255),   # Orange
    'Zone D': (20/255, 184/255, 166/255),   # Turquoise
    'Zone E': (236/255, 72/255, 153/255),   # Rose
    'Zone F': (99/255, 102/255, 241/255)    # Indigo
}

BASE_URL = 'https://gxomoissyprocedures.com'

def generate_qr_code(url, size=600):
    """Génère un QR Code et le sauvegarde temporairement"""
    qr = qrcode.QRCode(
        version=1,
        error_correction=qrcode.constants.ERROR_CORRECT_M,
        box_size=10,
        border=2,
    )
    qr.add_data(url)
    qr.make(fit=True)
    
    img = qr.make_image(fill_color="black", back_color="white")
    
    # Sauvegarder temporairement
    temp_path = f'/tmp/qr_temp_{datetime.now().timestamp()}.png'
    img.save(temp_path)
    return temp_path

def generate_pdf():
    """Génère le PDF avec tous les QR Codes"""
    filename = f'GXO-Moissy-QR-Codes-{datetime.now().strftime("%Y-%m-%d")}.pdf'
    output_path = f'/home/user/{filename}'
    
    print(f'🎨 Génération du PDF: {filename}')
    
    # Créer le PDF
    c = canvas.Canvas(output_path, pagesize=A4)
    page_width, page_height = A4
    
    # Dimensions (en mm converties en points)
    qr_width = 80 * mm
    qr_height = 80 * mm
    cols = 2
    margin = 10 * mm
    spacing = 10 * mm
    label_height = 20 * mm
    
    # En-tête première page
    c.setFont('Helvetica-Bold', 18)
    c.drawCentredString(page_width / 2, page_height - 40, 'QR Codes GXO Moissy - Scan Automatique')
    
    c.setFont('Helvetica', 11)
    date_str = datetime.now().strftime('%d/%m/%Y')
    c.drawCentredString(page_width / 2, page_height - 60, f'Généré le {date_str}')
    
    c.setFont('Helvetica', 9)
    c.drawCentredString(page_width / 2, page_height - 75, 
                       '45 QR Codes - Scanner avec votre smartphone pour démarrer le timer automatiquement')
    
    current_y = page_height - 100 * mm
    col = 0
    count = 0
    total_quais = sum(len(quais) for quais in QUAIS_CONFIG.values())
    
    for zone, quais in QUAIS_CONFIG.items():
        for quai_numero in quais:
            # Nouvelle page si nécessaire
            if current_y < margin + qr_height + label_height:
                c.showPage()
                current_y = page_height - margin
                col = 0
            
            x = margin + col * (qr_width + spacing)
            url = f'{BASE_URL}/scan?quai={quai_numero}'
            
            print(f'  📱 Génération QR Code Quai {quai_numero} ({zone})...')
            
            try:
                # Générer le QR Code
                qr_path = generate_qr_code(url)
                
                # Cadre
                c.setStrokeColorRGB(0.86, 0.86, 0.86)
                c.setLineWidth(0.5)
                c.rect(x, current_y, qr_width, qr_height + label_height)
                
                # Barre de couleur de zone
                color = ZONE_COLORS[zone]
                c.setFillColorRGB(*color)
                c.rect(x, current_y + qr_height + label_height - 8*mm, qr_width, 8*mm, fill=1)
                
                # Label zone
                c.setFillColorRGB(1, 1, 1)
                c.setFont('Helvetica-Bold', 10)
                c.drawCentredString(x + qr_width / 2, current_y + qr_height + label_height - 5*mm, zone)
                
                # QR Code
                c.drawImage(qr_path, x + 5*mm, current_y + 10*mm, qr_width - 10*mm, qr_width - 10*mm)
                
                # Supprimer le fichier temporaire
                os.remove(qr_path)
                
                # Label quai
                c.setFillColorRGB(0, 0, 0)
                c.setFont('Helvetica-Bold', 16)
                c.drawCentredString(x + qr_width / 2, current_y + 6*mm, f'Quai {quai_numero}')
                
                # URL
                c.setFont('Helvetica', 7)
                c.drawCentredString(x + qr_width / 2, current_y + 2*mm, url)
                
                # Instructions
                c.setFont('Helvetica', 6)
                c.setFillColorRGB(0.4, 0.4, 0.4)
                c.drawCentredString(x + qr_width / 2, current_y - 2*mm, 'Scanner pour démarrer le timer')
                
                count += 1
                col += 1
                
                if col >= cols:
                    col = 0
                    current_y -= (qr_height + label_height + spacing)
                    
            except Exception as e:
                print(f'❌ Erreur génération QR Code {quai_numero}: {e}')
    
    # Footer (sera ajouté sur la page courante uniquement)
    c.setFont('Helvetica', 8)
    c.setFillColorRGB(0.5, 0.5, 0.5)
    c.drawCentredString(page_width / 2, 15, 
                       f'GXO Moissy - {count} QR Codes Générés - {date_str}')
    
    # Sauvegarder le PDF
    c.save()
    
    print(f'\n✅ PDF généré avec succès!')
    print(f'📍 Emplacement: {output_path}')
    print(f'📊 Total: {count} QR Codes générés')
    
    return output_path

if __name__ == '__main__':
    try:
        pdf_path = generate_pdf()
        print(f'\n🎉 Génération terminée!')
        print(f'📄 Fichier disponible: {pdf_path}')
        print(f'\n💡 Commandes suivantes:')
        print(f'   - Télécharger: Utilisez l\'interface web ou SFTP')
        print(f'   - Imprimer: 300 DPI minimum, A4, couleur')
    except Exception as e:
        print(f'\n❌ Erreur: {e}')
        exit(1)

File: scripts/generate-qr-pdf.js
// Script Node.js pour générer le PDF des QR Codes
// Ce script génère un PDF avec tous les QR Codes des 45 quais GXO Moissy

const QRCode = require('qrcode');
const { jsPDF } = require('jspdf');
const fs = require('fs');
const path = require('path');

// Configuration des quais GXO Moissy
const QUAIS_CONFIG = {
  'Zone A': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
  'Zone B': [32, 33, 34, 35, 36, 37, 38],
  'Zone C': [45, 46, 47, 48, 49],
  'Zone D': [60, 61, 62, 67, 68, 69],
  'Zone E': [75, 76, 77, 78, 79, 81, 82, 83, 84, 85, 86, 87],
  'Zone F': [99, 100, 101, 102, 103]
};

const ZONE_COLORS = {
  'Zone A': { r: 59, g: 130, b: 246 },   // Bleu
  'Zone B': { r: 139, g: 92, b: 246 },   // Violet
  'Zone C': { r: 245, g: 158, b: 11 },   // Orange
  'Zone D': { r: 20, g: 184, b: 166 },   // Turquoise
  'Zone E': { r: 236, g: 72, b: 153 },   // Rose
  'Zone F': { r: 99, g: 102, b: 241 }    // Indigo
};

const BASE_URL = 'https://gxomoissyprocedures.com';

async function generateQRCodesPDF() {
  console.log('🎨 Génération du PDF des QR Codes GXO Moissy...');
  
  const pdf = new jsPDF('portrait', 'mm', 'a4');
  const pageWidth = pdf.internal.pageSize.getWidth();
  const pageHeight = pdf.internal.pageSize.getHeight();
  
  // Dimensions
  const qrWidth = 80;
  const qrHeight = 80;
  const cols = 2;
  const margin = 10;
  const spacing = 10;
  const labelHeight = 20;
  
  // En-tête première page
  pdf.setFont('helvetica', 'bold');
  pdf.setFontSize(18);
  pdf.text('QR Codes GXO Moissy - Scan Automatique', pageWidth / 2, 15, { align: 'center' });
  
  pdf.setFontSize(11);
  pdf.setFont('helvetica', 'normal');
  const date = new Date().toLocaleDateString('fr-FR');
  pdf.text(`Généré le ${date}`, pageWidth / 2, 22, { align: 'center' });
  
  pdf.setFontSize(9);
  pdf.text('45 QR Codes - Scanner avec votre smartphone pour démarrer le timer automatiquement', pageWidth / 2, 27, { align: 'center' });
  
  let currentY = 35;
  let col = 0;
  let count = 0;
  const totalQuais = Object.values(QUAIS_CONFIG).flat().length;
  
  for (const [zone, quais] of Object.entries(QUAIS_CONFIG)) {
    for (const quaiNumero of quais) {
      // Nouvelle page si nécessaire
      if (currentY + qrHeight + labelHeight > pageHeight - margin) {
        pdf.addPage();
        currentY = margin;
        col = 0;
      }
      
      const x = margin + col * (qrWidth + spacing);
      const url = `${BASE_URL}/scan?quai=${quaiNumero}`;
      
      console.log(`  📱 Génération QR Code Quai ${quaiNumero} (${zone})...`);
      
      try {
        // Générer le QR Code
        const qrDataURL = await QRCode.toDataURL(url, {
          width: 600,
          margin: 2,
          color: {
            dark: '#000000',
            light: '#FFFFFF'
          }
        });
        
        // Cadre
        pdf.setDrawColor(220);
        pdf.setLineWidth(0.5);
        pdf.rect(x, currentY, qrWidth, qrHeight + labelHeight);
        
        // Barre de couleur de zone
        const color = ZONE_COLORS[zone];
        pdf.setFillColor(color.r, color.g, color.b);
        pdf.rect(x, currentY, qrWidth, 8, 'F');
        
        // Label zone
        pdf.setTextColor(255, 255, 255);
        pdf.setFontSize(10);
        pdf.setFont('helvetica', 'bold');
        pdf.text(zone, x + qrWidth / 2, currentY + 5.5, { align: 'center' });
        
        // QR Code
        pdf.addImage(qrDataURL, 'PNG', x + 5, currentY + 10, qrWidth - 10, qrWidth - 10);
        
        // Label quai
        pdf.setTextColor(0, 0, 0);
        pdf.setFontSize(16);
        pdf.setFont('helvetica', 'bold');
        pdf.text(`Quai ${quaiNumero}`, x + qrWidth / 2, currentY + qrHeight + 6, { align: 'center' });
        
        // URL
        pdf.setFontSize(7);
        pdf.setFont('helvetica', 'normal');
        pdf.text(url, x + qrWidth / 2, currentY + qrHeight + 12, { 
          align: 'center',
          maxWidth: qrWidth - 4
        });
        
        // Instructions
        pdf.setFontSize(6);
        pdf.setTextColor(100, 100, 100);
        pdf.text('Scanner pour démarrer le timer', x + qrWidth / 2, currentY + qrHeight + 16, { align: 'center' });
        
        count++;
        col++;
        
        if (col >= cols) {
          col = 0;
          currentY += qrHeight + labelHeight + spacing;
        }
      } catch (error) {
        console.error(`❌ Erreur génération QR Code ${quaiNumero}:`, error);
      }
    }
  }
  
  // Footer sur toutes les pages
  const totalPages = pdf.internal.pages.length - 1;
  for (let i = 1; i <= totalPages; i++) {
    pdf.setPage(i);
    pdf.setFontSize(8);
    pdf.setTextColor(128);
    pdf.text(
      `Page ${i}/${totalPages} - GXO Moissy - ${count} QR Codes Générés`,
      pageWidth / 2,
      pageHeight - 5,
      { align: 'center' }
    );
  }
  
  // Sauvegarder
  const filename = `GXO-Moissy-QR-Codes-${new Date().toISOString().split('T')[0]}.pdf`;
  const outputPath = path.join(__dirname, filename);
  
  pdf.save(outputPath);
  
  console.log(`✅ PDF généré avec succès: ${filename}`);
  console.log(`📍 Emplacement: ${outputPath}`);
  console.log(`📊 Total: ${count} QR Codes générés`);
  
  return { filename, outputPath, count };
}

// Exécution
if (require.main === module) {
  generateQRCodesPDF()
    .then(result => {
      console.log('\n🎉 Génération terminée !');
      console.log(`📄 Fichier: ${result.filename}`);
      console.log(`📦 QR Codes: ${result.count}/45`);
    })
    .catch(error => {
      console.error('❌ Erreur:', error);
      process.exit(1);
    });
}

module.exports = { generateQRCodesPDF };
